core/src/main/java/feign/AsyncClient.java

```diff
@@ -0,0 +1,59 @@
+/**
+ * 
+ */
+package feign;
+
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Future;
+
+import feign.Request.Options;
+
+/**
+ * Submits HTTP {@link Request requests} asynchronously, with an optional context.
+ */
+public interface AsyncClient<C> {
+
+	/**
+	 * Executes the request asynchronously. Calling {@link CompletableFuture#cancel(boolean)} on the result
+	 * may cause the execution to be cancelled / aborted, but this is not guaranteed.
+	 * 
+	 * @param request safe to replay
+	 * @param options options to apply to this request
+	 * @param contextOpt - the optional context. The client should update this appropriately based
+	 * on the received response before completing the result.
+	 * @return a {@link CompletableFuture} to be completed with the response, or completed exceptionally
+	 * otherwise, for example with an {@link java.io.IOException} on a network error connecting to {@link Request#url()}.
+	 */
+	CompletableFuture<Response> execute(Request request, Options options, Optional<C> contextOpt);
```

Comment:
`contextOpt` is a confusing and ambiguous to me.  What is this context for?

Reply:
Generally a request happens within some context - for example, there may be session cookies or headers that need to be sent and maintained as part of the calls.

Reply:
Thanks for the explanation, can we rename it to `requestContext` to be more explicit?

Reply:
Done in latest commit